### PR TITLE
Sentinel 2 S3 provider

### PR DESCRIPTION
**Pull request by michaelmattig**

### CODE REVIEW

**Comment by michaelmattig** [comment]
[Externe Datenquelle für die S2 S3 Tiles bauen mit earth search.](https://trello.com/c/4xuNOFJn/201-externe-datenquelle-f%C3%BCr-die-s2-s3-tiles-bauen-mit-earth-search)

File: operators/src/error.rs [hunk 1]
```diff
@@ -193,6 +193,12 @@ pub enum Error {
 
     TemporalRasterAggregationLastValidRequiresNoData,
     TemporalRasterAggregationFirstValidRequiresNoData,
+
+    LoadingInfo {
+        reason: String, // TODO: find cleaner way to make errors in services crate visible
```
Comment (jdroenner) [inline operators/src/error.rs:198]: maybe just a box error?

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
+    async fn initialize(
+        self: Box<Self>,
+    ) -> crate::error::Result<Box<dyn crate::datasets::listing::DatasetProvider>> {
+        Ok(Box::new(SentinelS2L2aCogsDataProvider::new(
+            self.id,
+            self.api_url,
+        )))
+    }
+
+    fn type_name(&self) -> String {
```
Comment (jdroenner) [inline services/src/datasets/external/sentinel_s2_l2a_cogs.rs:55]: whats this? / why is this needed? why is it not static?

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+
+#[derive(Debug, Clone)]
+pub struct SentinelDataset {
+    band: Band,
+    zone: Zone,
+    listing: DatasetListing,
+}
+
+pub struct SentinelS2L2aCogsDataProvider {
+    #[allow(dead_code)]
```
Comment (jdroenner) [inline services/src/datasets/external/sentinel_s2_l2a_cogs.rs:111]: why is this dead?

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs [hunk 4]
```diff
@@ -0,0 +1,640 @@
...
+            id,
+            api_url,
+            datasets: Self::create_datasets(&id, &meta_data),
+            meta_data,
+        }
+    }
+
+    fn load_metadata() -> SentinelMetaData {
+        // TODO: fetch dataset metadata from config or remote
+        // TODO: is there a no data value?
```
Comment (jdroenner) [inline services/src/datasets/external/sentinel_s2_l2a_cogs.rs:132]: yes and you also set it :)

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs [hunk 5]
```diff
@@ -0,0 +1,640 @@
...
+        SentinelMetaData {
```
Comment (jdroenner) [inline services/src/datasets/external/sentinel_s2_l2a_cogs.rs:133]: this will never scale there are like 12 bands and 80 zones.

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs [hunk 6]
```diff
@@ -0,0 +1,640 @@
...
+                })
+            })
+            .collect()
+    }
+}
+
+#[async_trait]
+impl DatasetProvider for SentinelS2L2aCogsDataProvider {
+    async fn list(&self, _options: Validated<DatasetListOptions>) -> Result<Vec<DatasetListing>> {
+        // TODO: options
```
Comment (jdroenner) [inline services/src/datasets/external/sentinel_s2_l2a_cogs.rs:194]: what kind of options?

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs [hunk 7]
```diff
@@ -0,0 +1,640 @@
...
+    ) -> Result<Vec<StacFeature>> {
+        let mut features = vec![];
+
+        let mut collection = self.load_collection(params, 1).await?;
+        features.append(&mut collection.features);
+
+        let num_pages =
+            (collection.context.matched as f64 / collection.context.limit as f64).ceil() as u32;
+
+        for page in 2..=num_pages {
```
Comment (jdroenner) [inline services/src/datasets/external/sentinel_s2_l2a_cogs.rs:344]: what happens if num_pages is 1?

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs [hunk 8]
```diff
@@ -0,0 +1,640 @@
...
+            time.start()
+                .as_utc_date_time()
+                .ok_or(geoengine_operators::error::Error::DataType {
+                    source: geoengine_datatypes::error::Error::NoDateTimeValid {
+                        time_instance: time.start(),
+                    },
+                })?;
+
+        // shift start by 1 day to ensure getting the most recent data for start time
+        let t_start = t_start - Duration::days(1);
```
Comment (jdroenner) [inline services/src/datasets/external/sentinel_s2_l2a_cogs.rs:387]: i guess we can also move it by -1 minutes. this will be enough to include a tile if it started before.

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs [hunk 9]
```diff
@@ -0,0 +1,640 @@
...
+        debug!("number of features in current zone: {}", num_features);
+        for i in 0..num_features {
+            let feature = &features[i];
+
+            let start = feature.properties.datetime;
+            // feature is valid until next feature starts
+            let end = if i < num_features - 1 {
+                features[i + 1].properties.datetime
+            } else {
+                start + Duration::hours(12) // TODO: determine correct validity for last tile
```
Comment (jdroenner) [inline services/src/datasets/external/sentinel_s2_l2a_cogs.rs:241]: maybe just add 1 minute. this should be fine..

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs [hunk 10]
```diff
@@ -0,0 +1,640 @@
...
+
+    async fn result_descriptor(&self) -> geoengine_operators::util::Result<RasterResultDescriptor> {
+        Ok(RasterResultDescriptor {
+            data_type: self.band.data_type,
+            spatial_reference: SpatialReference::new(
+                SpatialReferenceAuthority::Epsg,
+                self.zone.epsg,
+            )
+            .into(),
+            measurement: Measurement::Unitless,
```
Comment (jdroenner) [inline services/src/datasets/external/sentinel_s2_l2a_cogs.rs:424]: add a todo to add this later

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs [hunk 11]
```diff
@@ -0,0 +1,640 @@
...
+            no_data_value: self.band.no_data_value,
+        })
+    }
+
+    fn box_clone(&self) -> Box<dyn MetaData<GdalLoadingInfo, RasterResultDescriptor>> {
```
Comment (jdroenner) [inline services/src/datasets/external/sentinel_s2_l2a_cogs.rs:429]: can we default impl this?

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs [hunk 12]
```diff
@@ -0,0 +1,640 @@
...
+        Ok(Box::new(SentinelS2L2aCogsMetaData {
+            api_url: self.api_url.clone(),
+            zone: dataset.zone.clone(),
+            band: dataset.band.clone(),
+        }))
+    }
+}
+
+#[async_trait]
+impl MetaDataProvider<MockDatasetDataSourceLoadingInfo, VectorResultDescriptor>
```
Comment (jdroenner) [inline services/src/datasets/external/sentinel_s2_l2a_cogs.rs:457]: do we really have to implement all kinds if they are not really implemented? There should be a task to solve this

**Review by jdroenner — COMMENTED** [review]
i think this works for a first version. please explain some design choices

File: operators/src/error.rs (hunk 1 above)
Comment (michaelmattig) [reply operators/src/error.rs:198]: the error occurs in the services crate so the operators Error cannot know the error and thus also a `Box<geoengine_services::Error>` won't work

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 2 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:55]: it is the type of the provider, while the name ist the concrete instance. "WCS" vs. "DWD Geoserver".
It cannot be static because then it wouldn't be deserializable with type_tag

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 3 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:111]: it's dead right now because I create all the MetaData on `new` but in practice it should be retrieved from STAC and in order to generate the external dataset id the provider needs to know its own provider id, which is part of the dataset id. I can remove it for now if you want

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs [hunk 13]
```diff
@@ -0,0 +1,640 @@
...
+    band: Band,
+    zone: Zone,
+    listing: DatasetListing,
+}
+
+pub struct SentinelS2L2aCogsDataProvider {
+    #[allow(dead_code)]
+    id: DatasetProviderId,
+    api_url: String,
+    #[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:114]: cf. above

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 5 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:133]: yeah, we would need hierarchical datasets here I guess

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 6 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:194]: the listing options. sorting, filtering etc.

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 7 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:344]: then no further pages have to be downloaded. the first page is always downloaded

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 8 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:387]: are you sure? is the satellite also recording while it's on the night side?

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 9 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:241]: are you sure? is the satellite also recording while it's on the night side?

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 12 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:457]: a case for trait specialization... maybe we can create a macro that builds NotImplented-stubs

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 11 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:428]: I guess we could if we create another trait CloneableMetaData that has only that function. Shall I try it?

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 4 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs]: 3bbd917d73e944b7697dfdd5f375cf69d7c52ffb

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 10 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:423]: 3bbd917d73e944b7697dfdd5f375cf69d7c52ffb

File: operators/src/error.rs (hunk 1 above)
Comment (jdroenner) [reply operators/src/error.rs:198]: maybe use Trait std::error::Error or Into Trait std::error::Error

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 3 above)
Comment (jdroenner) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:111]: add a todo

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (jdroenner) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:114]: as above

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 5 above)
Comment (jdroenner) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:132]: or a different selection ui

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 8 above)
Comment (jdroenner) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:386]: no but then there is no tile at night. i think tats a better semantic then a tile which is for ever valid

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 9 above)
Comment (jdroenner) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:240]: as above

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 11 above)
Comment (jdroenner) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:428]: no then it is to much overhead. cant you simply implement it in the MetaData trait?

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 12 above)
Comment (jdroenner) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:456]: i'm still not convinced that we need to require all the implementations everyehere. This is what happens if you do this. i guess there should be a better metadata provider instantiation / retrival mechanisim

File: operators/src/error.rs (hunk 1 above)
Comment (michaelmattig) [reply operators/src/error.rs:198]: `std::error::Error` is not Send + Sync und thus our `Result` would no longer be Send + Sync. Also our own `Error` doesn't even implement `std::error::Error` or does it?

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 3 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs]: 3ed69c0b2d4e186ec765d553ddc10d46fc780e39

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs]: 3ed69c0b2d4e186ec765d553ddc10d46fc780e39

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 5 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:127]: we could have a special dialog for Sentinal datasets but I would prefer a generic solution

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 8 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:381]: if I only shift by 1 minute there may be no tile at all because the latest tile was from the prevous day. ThatÄs what I menat.

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 9 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:235]: if I only shift by one minute there may be no tile at all because the latest tile is from previous day

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 11 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:423]: I don't know how. I used explicitly this workaround of having a box_clone method and implementing Clone for the Box because it's not possible IMHO to directly make the trait Clone

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 12 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:451]: I can't see how that's possible, you would have to explain me how to

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 5 above)
Comment (jdroenner) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:127]: so add a task for it

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 8 above)
Comment (jdroenner) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:381]: yes but i think this is ok.  ypu need to aggregate anyway

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 9 above)
Comment (jdroenner) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:235]: yes but i think this is ok. ypu need to aggregate anyway

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 11 above)
Comment (jdroenner) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:423]: then we have to keep it like this

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 12 above)
Comment (jdroenner) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:451]: yes but that is a another task

File: operators/src/error.rs (hunk 1 above)
Comment (jdroenner) [reply operators/src/error.rs:198]: you can do:

```
#[snafu(display("LoadingInfoError: {}", source))]
    LoadingInfo {
        source: Box<dyn std::error::Error + Send + Sync>, // TODO: find cleaner way to make errors in services crate visible
    },

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 5 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs:127]: we already have one https://trello.com/c/ebe6iQwz/115-bring-back-datasets-with-more-then-one-layer-band

File: operators/src/error.rs (hunk 1 above)
Comment (michaelmattig) [reply operators/src/error.rs]: 4e19dc43043d20988a199028294ebe05e3ea7533

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 8 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs]: 89397a1bc02a646b760c2b2655eec984131b749b

File: services/src/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 9 above)
Comment (michaelmattig) [reply services/src/datasets/external/sentinel_s2_l2a_cogs.rs]: 89397a1bc02a646b760c2b2655eec984131b749b
